File: day-12-using-mcp-server-tool/SimpleMCPServer/src/main/kotlin/com/example/mcp/mcp/McpProtocol.kt
package com.example.mcp.mcp

import com.example.mcp.plugins.*
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*

/**
 * MCP Protocol version
 */
const val MCP_PROTOCOL_VERSION = "2024-11-05"

/**
 * JSON-RPC request
 */
@Serializable
data class JsonRpcRequest(
    val jsonrpc: String = "2.0",
    val id: JsonElement? = null,
    val method: String,
    val params: JsonObject? = null
)

/**
 * JSON-RPC response
 */
@Serializable
data class JsonRpcResponse(
    val jsonrpc: String = "2.0",
    val id: JsonElement? = null,
    val result: JsonElement? = null,
    val error: JsonRpcError? = null
)

@Serializable
data class JsonRpcError(
    val code: Int,
    val message: String,
    val data: JsonElement? = null
)

/**
 * MCP Server capabilities
 */
@Serializable
data class ServerCapabilities(
    val tools: ToolsCapability? = ToolsCapability()
)

@Serializable
data class ToolsCapability(
    val listChanged: Boolean = false
)

/**
 * MCP Server info
 */
@Serializable
data class ServerInfo(
    val name: String,
    val version: String
)

/**
 * Initialize result
 */
@Serializable
data class InitializeResult(
    val protocolVersion: String,
    val capabilities: ServerCapabilities,
    val serverInfo: ServerInfo
)

/**
 * Few-shot example for serialization
 */
@Serializable
data class FewShotExampleDto(
    val request: String,
    val params: JsonObject
)

/**
 * Tool definition for listing
 */
@Serializable
data class ToolDefinition(
    val name: String,
    val description: String,
    val inputSchema: JsonObject,
    val fewShotExamples: List<FewShotExampleDto>? = null
)

/**
 * Tools list result
 */
@Serializable
data class ToolsListResult(
    val tools: List<ToolDefinition>
)

/**
 * Tool call result content
 */
@Serializable
data class ToolResultContent(
    val type: String,
    val text: String? = null,
    val data: JsonElement? = null
)

/**
 * Tool call result
 */
@Serializable
data class CallToolResult(
    val content: List<ToolResultContent>,
    val isError: Boolean = false
)

/**
 * MCP Protocol handler
 */
class McpProtocolHandler(
    private val toolRegistry: ToolRegistry,
    private val serverName: String = "mcp-server",
    private val serverVersion: String = "1.0.0"
) {
    private val json = Json {
        ignoreUnknownKeys = true
        encodeDefaults = true
    }

    suspend fun handleRequest(requestJson: String): String {
        val request = try {
            json.decodeFromString<JsonRpcRequest>(requestJson)
        } catch (e: Exception) {
            return json.encodeToString(
                JsonRpcResponse.serializer(),
                JsonRpcResponse(
                    error = JsonRpcError(
                        code = -32700,
                        message = "Parse error: ${e.message}"
                    )
                )
            )
        }

        val result = when (request.method) {
            "initialize" -> handleInitialize(request)
            "initialized" -> handleInitialized(request)
            "tools/list" -> handleToolsList(request)
            "tools/call" -> handleToolsCall(request)
            "ping" -> handlePing(request)
            else -> JsonRpcResponse(
                id = request.id,
                error = JsonRpcError(
                    code = -32601,
                    message = "Method not found: ${request.method}"
                )
            )
        }

        return json.encodeToString(JsonRpcResponse.serializer(), result)
    }

    private fun handleInitialize(request: JsonRpcRequest): JsonRpcResponse {
        val initResult = InitializeResult(
            protocolVersion = MCP_PROTOCOL_VERSION,
            capabilities = ServerCapabilities(
                tools = ToolsCapability(listChanged = false)
            ),
            serverInfo = ServerInfo(
                name = serverName,
                version = serverVersion
            )
        )

        return JsonRpcResponse(
            id = request.id,
            result = json.encodeToJsonElement(InitializeResult.serializer(), initResult)
        )
    }

    private fun handleInitialized(request: JsonRpcRequest): JsonRpcResponse {
        // Notification, no response needed but we'll acknowledge
        return JsonRpcResponse(
            id = request.id,
            result = JsonObject(emptyMap())
        )
    }

    private fun handleToolsList(request: JsonRpcRequest): JsonRpcResponse {
        val tools = toolRegistry.getAllTools().map { tool ->
            val examples = tool.fewShotExamples.takeIf { it.isNotEmpty() }?.map { ex ->
                FewShotExampleDto(request = ex.request, params = ex.params)
            }
            ToolDefinition(
                name = tool.name,
                description = tool.description,
                inputSchema = schemaToJsonObject(tool.inputSchema),
                fewShotExamples = examples
            )
        }

        val result = ToolsListResult(tools = tools)

        return JsonRpcResponse(
            id = request.id,
            result = json.encodeToJsonElement(ToolsListResult.serializer(), result)
        )
    }

    private suspend fun handleToolsCall(request: JsonRpcRequest): JsonRpcResponse {
        val params = request.params ?: return JsonRpcResponse(
            id = request.id,
            error = JsonRpcError(
                code = -32602,
                message = "Invalid params: missing parameters"
            )
        )

        val toolName = params["name"]?.jsonPrimitive?.contentOrNull ?: return JsonRpcResponse(
            id = request.id,
            error = JsonRpcError(
                code = -32602,
                message = "Invalid params: missing tool name"
            )
        )

        val arguments = params["arguments"]?.jsonObject ?: JsonObject(emptyMap())

        val tool = toolRegistry.getTool(toolName) ?: return JsonRpcResponse(
            id = request.id,
            error = JsonRpcError(
                code = -32602,
                message = "Tool not found: $toolName"
            )
        )

        return try {
            val toolResult = tool.execute(arguments)
            val content = when (toolResult) {
                is ToolResult.Success -> {
                    CallToolResult(
                        content = toolResult.content.map { c ->
                            when (c) {
                                is ToolContent.Text -> ToolResultContent(type = "text", text = c.text)
                                is ToolContent.Json -> ToolResultContent(type = "json", data = c.data)
                            }
                        },
                        isError = false
                    )
                }
                is ToolResult.Error -> {
                    CallToolResult(
                        content = listOf(ToolResultContent(type = "text", text = toolResult.message)),
                        isError = true
                    )
                }
            }

            JsonRpcResponse(
                id = request.id,
                result = json.encodeToJsonElement(CallToolResult.serializer(), content)
            )
        } catch (e: Exception) {
            JsonRpcResponse(
                id = request.id,
                error = JsonRpcError(
                    code = -32000,
                    message = "Tool execution error: ${e.message}"
                )
            )
        }
    }

    private fun handlePing(request: JsonRpcRequest): JsonRpcResponse {
        return JsonRpcResponse(
            id = request.id,
            result = JsonObject(emptyMap())
        )
    }

    private fun schemaToJsonObject(schema: JsonSchema): JsonObject {
        val properties = buildJsonObject {
            for ((name, prop) in schema.properties) {
                put(name, buildJsonObject {
                    put("type", JsonPrimitive(prop.type))
                    prop.description?.let { put("description", JsonPrimitive(it)) }
                    prop.enum?.let { enumList ->
                        put("enum", JsonArray(enumList.map { JsonPrimitive(it) }))
                    }
                    prop.default?.let { put("default", it) }
                })
            }
        }

        return buildJsonObject {
            put("type", JsonPrimitive(schema.type))
            put("properties", properties)
            if (schema.required.isNotEmpty()) {
                put("required", JsonArray(schema.required.map { JsonPrimitive(it) }))
            }
            schema.description?.let { put("description", JsonPrimitive(it)) }
        }
    }
}

File: day-15-environment/mcp-servers/shared/mcp_protocol.py
"""MCP (Model Context Protocol) JSON-RPC 2.0 handler.

Implements the server side of the MCP protocol:
- initialize / initialized handshake
- tools/list — advertise available tools
- tools/call — execute a specific tool
- ping — keepalive check

All communication uses JSON-RPC 2.0 framing over an SSE transport.
"""

import json
import logging
from typing import Any

logger = logging.getLogger(__name__)

MCP_PROTOCOL_VERSION = "2024-11-05"


def _response(request_id: Any, result: Any = None, error: dict = None) -> dict:
    """Build a JSON-RPC 2.0 response object."""
    resp = {"jsonrpc": "2.0", "id": request_id}
    if error:
        resp["error"] = error
    else:
        resp["result"] = result
    return resp


class McpProtocolHandler:
    """Stateless handler for incoming MCP JSON-RPC requests."""

    def __init__(self, tools: list, server_name: str, server_version: str = "1.0.0"):
        self.tools = {t.name: t for t in tools}
        self.server_name = server_name
        self.server_version = server_version

    async def handle_request(self, request_json: str) -> str:
        """Parse and dispatch a single JSON-RPC request. Returns a JSON string."""
        try:
            request = json.loads(request_json)
        except json.JSONDecodeError as e:
            return json.dumps(_response(None, error={"code": -32700, "message": f"Parse error: {e}"}))

        request_id = request.get("id")
        method = request.get("method", "")
        params = request.get("params", {})

        if method == "initialize":
            result = self._handle_initialize()
        elif method == "initialized":
            result = {}
        elif method == "tools/list":
            result = self._handle_tools_list()
        elif method == "tools/call":
            return await self._handle_tools_call(request_id, params)
        elif method == "ping":
            result = {}
        else:
            return json.dumps(_response(request_id, error={
                "code": -32601,
                "message": f"Method not found: {method}",
            }))

        return json.dumps(_response(request_id, result=result))

    def _handle_initialize(self) -> dict:
        return {
            "protocolVersion": MCP_PROTOCOL_VERSION,
            "capabilities": {
                "tools": {"listChanged": False},
            },
            "serverInfo": {
                "name": self.server_name,
                "version": self.server_version,
            },
        }

    def _handle_tools_list(self) -> dict:
        tools_list = []
        for tool in self.tools.values():
            tool_info = {
                "name": tool.name,
                "description": tool.description,
                "inputSchema": tool.input_schema,
            }
            # Include fewShotExamples if available (support both snake_case and camelCase)
            few_shot = getattr(tool, 'few_shot_examples', None) or getattr(tool, 'fewShotExamples', None)
            if few_shot:
                tool_info["fewShotExamples"] = few_shot
            # Include negativeFewShotExamples if available (examples when NOT to use this tool)
            negative_few_shot = getattr(tool, 'negativeFewShotExamples', None)
            if negative_few_shot:
                tool_info["negativeFewShotExamples"] = negative_few_shot
            tools_list.append(tool_info)
        return {"tools": tools_list}

    async def _handle_tools_call(self, request_id: Any, params: dict) -> str:
        tool_name = params.get("name")
        if not tool_name:
            return json.dumps(_response(request_id, error={
                "code": -32602,
                "message": "Invalid params: missing tool name",
            }))

        tool = self.tools.get(tool_name)
        if not tool:
            return json.dumps(_response(request_id, error={
                "code": -32602,
                "message": f"Tool not found: {tool_name}",
            }))

        arguments = params.get("arguments", {})
        logger.info(f"Tool call: {tool_name} | arguments: {json.dumps(arguments)}")

        try:
            tool_result = await tool.execute(arguments)
            result = {
                "content": [{"type": "text", "text": tool_result.content}],
                "isError": tool_result.is_error,
            }
            return json.dumps(_response(request_id, result=result))
        except Exception as e:
            logger.error(f"Tool execution error ({tool_name}): {e}")
            return json.dumps(_response(request_id, error={
                "code": -32000,
                "message": f"Tool execution error: {e}",
            }))
